File: Nodal_Discontinuous_Galerkin_Methods_Hesthaven/Codes1.1/CFD2D/CurvedINSPressureSetUp2D.m
function [PRperm, PRsystemC, PRsystemCT, rhsbcPR] = CurvedINSPressureSetUp2D(dt, nu, BCfunction)

% function [PRperm, PRsystemC, PRsystemCT, rhsbcPR] = CurvedINSPressureSetUp2D(dt, nu, BCfunction)
% Purpose: build pressure system and boundary forcing term

Globals2D;

% choose order to integrate exactly
Nint = ceil(3*N/2);

% build cubature nodes for all elements
CubatureOrder = 2*(Nint+1); cub = CubatureVolumeMesh2D(CubatureOrder);

% build Gauss node data for all element faces
NGauss = (Nint+1); gauss = GaussFaceMesh2D(NGauss);

% Convert boundary conditions to Dirichlet & Neumann
saveBCType = BCType;
ids = find(saveBCType==Out); BCType(ids) = Dirichlet;
ids = find(saveBCType==In | saveBCType==Wall | saveBCType==Cyl);  
BCType(ids) = Neuman;

% Form inhomogeneous boundary term for rhs data 
[bcUx,bcUy,bcPR] = ...
feval(BCfunction, gauss.x, gauss.y, gauss.nx, gauss.ny,...
      gauss.mapI, gauss.mapO, gauss.mapW, gauss.mapC, 0, nu);  

% Build pressure boundary condition forcing vector
[PRsystemBC] = CurvedPoissonIPDGbc2D();
rhsbcPR = PRsystemBC*bcPR(:);

% Build pressure system (all Neumann, excluding outflow)
[PRsystem, mm] = CurvedPoissonIPDG2D();

% Restore original boundary types
BCType = saveBCType;

% Find Reverse-Cuthill-Mckee ordering to reduce bandwith of Pressure matrix
PRperm     = symrcm(PRsystem);

% Apply row and column permutation to Pressure matrix
PRsystem   = PRsystem(PRperm, PRperm);

% Compute Cholesky factorization of Pressure matrix
PRsystemC  = chol(PRsystem); PRsystemCT = transpose(PRsystemC);
return;

File: Nodal_Discontinuous_Galerkin_Methods_Hesthaven/nudg++/trunk/Src/Examples2D/CurvedINS2D/CurvedINSPressureSetUp2D.cpp
// CurvedINSPressureSetUp2D.m
// function [PRperm, PRsystemC, PRsystemCT, rhsbcPR] = ...
//            CurvedINSPressureSetUp2D(dt, nu, BCfunction)
// 2007/06/10
//---------------------------------------------------------
#include "NDGLib_headers.h"
#include "CurvedINS2D.h"


//---------------------------------------------------------
void CurvedINS2D::CurvedINSPressureSetUp2D()
//---------------------------------------------------------
{
  //---------------------------
  double t1 = timer.read();
  //---------------------------

  // Note: assumes {m_cub,m_gauss} have been defined
  Cub2D&   cub = this->m_cub;
  Gauss2D& gauss = this->m_gauss;

  // function [PRperm, PRsystemC, PRsystemCT, rhsbcPR] = ...
  //   CurvedINSPressureSetUp2D(dt, nu, BCfunction)
  //
  // Purpose: build pressure system and boundary forcing term

  CSd *PRsystemBC = new CSd("PRbc");  // NBN: delete before chol()
  CSd *mm = new CSd("mmP");           // NBN: delete before chol()
  CSd PRsystem("PR"); IVec ids;       // NBN: passed to chol()

  // save original boundary types
  saveBCType = BCType;

  // Convert {Out} boundary conditions to Dirichlet
  ids = find(saveBCType, '=', (int)BC_Out);   BCType(ids) = BC_Dirichlet;
  
  // Convert {In,Wall,Cyl} boundary conditions to Neumann
  ids = find(saveBCType, '=', (int)BC_In );   BCType(ids) = BC_Neuman;
  ids = find(saveBCType, '=', (int)BC_Wall);  BCType(ids) = BC_Neuman;
  ids = find(saveBCType, '=', (int)BC_Cyl);   BCType(ids) = BC_Neuman;

  // Check: dumpIMat(BCType, "Pressure BCType");

  // Form inhomogeneous boundary term for rhs data
  (this->*ExactSolutionBC)(gauss.x, gauss.y, gauss.nx, gauss.ny, 
             gauss.mapI, gauss.mapO, gauss.mapW, gauss.mapC, 0.0, nu,
             bcUx, bcUy, bcPR, bcdUndt);

  // Build pressure boundary condition forcing vector
  CurvedPoissonIPDGbc2D(m_gauss, (*PRsystemBC));
  refrhsbcPR = (*PRsystemBC) * bcPR;
  delete PRsystemBC; PRsystemBC=NULL;

  // Build pressure system (all Neumann, excluding outflow)
  CurvedPoissonIPDG2D(m_gauss, m_cub, PRsystem, (*mm));
  delete mm; mm=NULL;


#if (0)
  // check against Matlab
  FILE* fp = fopen("nnP.dat", "w");
  PRsystem.write_ML(fp); fclose(fp);
#endif

  //-------------------------------------
  // factor Pressure Op
  //-------------------------------------
  PRsystemC->chol(PRsystem, 4);   // 4=CS_Chol option

  PRsystem.reset();               // force immediate deallocation
  BCType = saveBCType;            // Restore original boundary types

  //---------------------------
  time_setup += timer.read() - t1;
  //---------------------------
}
